src/persist-client/src/impl/machine.rs

```diff
@@ -363,7 +370,11 @@ where
         loop {
             match self.apply_unbatched_cmd(cmd, &mut work_fn).await {
                 Ok((seqno, x)) => match x {
-                    Ok(x) => return (seqno, x),
+                    Ok(x) => {
+                        // TODO: This is not a great place to hook this in.
+                        self.maybe_expire_readers();
+                        return (seqno, x);
```

Comment:
is there any risk in returning an older seqno than what it actually is? if `expire_readers` completed before this returns, assertions like https://github.com/MaterializeInc/materialize/blob/fec38f5b2b6c9ccef398f3d8e11c742e7b7f78b1/src/persist-client/src/impl/machine.rs#L117 will fail. 

that said, it's a debug assertion so I'm assuming it's not so critical that we actually want to panic (and if we did panic, that would also seems OK since this whole point of the PR is to make restarts safe 😁)